### PR TITLE
Add util methods for signing and signature validation

### PR DESCRIPTION
**Pull request by sadilchamishka**
### Proposed changes in this pull request

$subject

### CODE REVIEW

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java [hunk 1]
```diff
@@ -1962,4 +1968,70 @@ public static boolean isSCIM2UserMaxItemsPerPageEnabled() {
...
+     * @param signature    The signature to be verified.
+     * @param tenantDomain The tenant domain to which the data belongs.
+     * @return true if the signature is valid, false otherwise.
+     * @throws IdentityKeyStoreResolverException If an error occurs during the signature validation process.
+     */
+    public static boolean validateSignature(String data, byte[] signature, String tenantDomain)
+            throws IdentityKeyStoreResolverException {
+
+        try {
+            int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
```
Comment (hwupathum) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java:1985]: We can take this out of the try block

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java [hunk 2]
```diff
@@ -1962,4 +1968,70 @@ public static boolean isSCIM2UserMaxItemsPerPageEnabled() {
...
+
+        try {
+            if (MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
+                privateKey = keyStoreManager.getDefaultPrivateKey();
+            } else {
+                String tenantKeyStoreName = IdentityKeyStoreResolverUtil.buildTenantKeyStoreName(tenantDomain);
+                try {
+                    IdentityTenantUtil.initializeRegistry(tenantId);
+                } catch (IdentityException e) {
+                    throw new IdentityKeyStoreResolverException(
```
Comment (hwupathum) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java:2020]: The exception thrown here is caught by the outer catch block.

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java [hunk 3]
```diff
@@ -1962,4 +1968,70 @@ public static boolean isSCIM2UserMaxItemsPerPageEnabled() {
...
+
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+        KeyStoreManager keyStoreManager = KeyStoreManager.getInstance(tenantId);
+        PrivateKey privateKey;
+
+        try {
+            if (MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
+                privateKey = keyStoreManager.getDefaultPrivateKey();
+            } else {
+                String tenantKeyStoreName = IdentityKeyStoreResolverUtil.buildTenantKeyStoreName(tenantDomain);
```
Comment (hwupathum) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java:2016]: We can move the `catch (Exception e)` to here

**Comment by hwupathum** [comment]
Shall we rename the methods to reflect they are tenanted methods? Eg.
- validateTenantSignature (for validation)
- signWithTenantKey (for signing)

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java (hunk 1 above)
Comment (sadilchamishka) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java]: Fixed by - https://github.com/wso2/carbon-identity-framework/pull/6176/commits/43523e575a591197e6f3b04210f9be0a83ee4fe0

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java (hunk 3 above)
Comment (sadilchamishka) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java]: A different change is sent from this commit - https://github.com/wso2/carbon-identity-framework/pull/6176/commits/43523e575a591197e6f3b04210f9be0a83ee4fe0

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java (hunk 2 above)
Comment (sadilchamishka) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java]: Fixed by - https://github.com/wso2/carbon-identity-framework/pull/6176/commits/43523e575a591197e6f3b04210f9be0a83ee4fe0

**Comment by sadilchamishka** [comment]
> * validateTenantSignature

Refactored as `validateSignatureFromTenant`

**Review by hwupathum — DISMISSED** [review]

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/IdentityKeyStoreResolver.java [hunk 4]
```diff
@@ -297,9 +297,7 @@ public Certificate getCertificate(String tenantDomain, InboundProtocol inboundPr
                     String.format(ErrorMessages.ERROR_CODE_INVALID_ARGUMENT.getDescription(), "Tenant domain"));
         }
         if (inboundProtocol == null) {
-            throw new IdentityKeyStoreResolverException(
-                    ErrorMessages.ERROR_CODE_INVALID_ARGUMENT.getCode(),
-                    String.format(ErrorMessages.ERROR_CODE_INVALID_ARGUMENT.getDescription(), "Inbound protocol"));
+            return getCertificate(tenantDomain);
```
Comment (sadilchamishka) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/IdentityKeyStoreResolver.java:300]: As per the comments there is no need to throw client error, we can fallback to default by getting certificate of the tenant domain without bound to the protocol

**Review by indeewari — APPROVED** [review]

**Pull request merged**